woeip/apps/database/models.py

```diff
@@ -0,0 +1,118 @@
+
+    def __str__(self):
+        return f"{self.name} <{self.email}>"
+
+
+class RoleEnum(models.Model):
+    """Create, read, update, delete permissions"""
+    name = models.CharField(max_length=256)
+    can_create = models.BooleanField()
+    can_read = models.BooleanField()
+    can_update = models.BooleanField()
+    can_delete = models.BooleanField()
+
+    def __str__(self):
+        crud = []
+        if self.can_create:
+            crud.append('create')
+        if self.can_read:
+            crud.append('read')
+        if self.can_update:
+            crud.append('update')
+        if self.can_delete:
+            crud.append('delete')
+
+        crud = '/'.join(crud)
+
+        return f"{self.name} ({crud})"
+
+
+class Participant(models.Model):
+    name = models.CharField(max_length=256)
+    email = models.CharField(max_length=256)
+    organization = models.ForeignKey(Organization, on_delete=models.SET_NULL, blank=True, null=True)
+    role = models.ForeignKey(RoleEnum, on_delete=models.SET_NULL, blank=True, null=True)
+
+    def __str__(self):
+        return f"{self.name} <{self.email}> {self.organization} {self.role}"
+
+
+class RouteEnum(models.Model):
+    """Named routes
+    """
+    name = models.CharField(max_length=256)
+    latlon = LineStringField()
+
+    def __str__(self):
+        return f"{self.name}"
+
+
+class Session(models.Model):
+    """A single air quality outing. Can link to several SessionData, e.g., raw data files."""
+    date_collected = models.DateTimeField(auto_now_add=True)
+    route = models.ForeignKey(RouteEnum, on_delete=models.SET_NULL, blank=True, null=True)
+    collected_by = models.ForeignKey(Participant, on_delete=models.CASCADE)
+    metadata = JSONField(blank=True, null=True)
+
+    def __str__(self):
+        return f"{self.date_collected} {self.collected_by}"
+
+
+class Device(models.Model):
+    name = models.CharField(max_length=256)
+    manufacturer = models.CharField(max_length=256)
+    serial_number = models.CharField(max_length=256)
+    model_number = models.CharField(max_length=256)
+    calibration_date = models.DateField()
+    firmware_version = models.CharField(max_length=256)
+
+    def __str__(self):
+        return f"{self.name} {self.model_number} {self.serial_number}"
+
+
+class Sensor(models.Model):
+    """A sensor is something that measures something, i.e., it produces a single measurement value
+    at a time.
+    """
+    name = models.CharField(max_length=256)
+    unit = models.CharField(max_length=256)
+    device = models.ForeignKey(Device, on_delete=models.CASCADE)
+
+    def __str__(self):
+        return f"{self.name} ({self.device.name})"
+
+
+class SessionData(models.Model):
+    """The raw data file generated during a session. Assumes one and only one file per sensor,
+    although multiple sensors can be linked to one session"""
+    uri = models.CharField(max_length=256)
```

Comment:
What is this?

Reply:
By "this" do you mean the ``SessionData`` model? ``SessionData`` is the bridge between unprocessed raw data files and the preprocessed ``Data`` table. Nesting it this way means we can make ``Data`` have as few fields as possible -- the value, time, location, and a link to the session it came from. I considered it important to keep the schema for ``Data`` lean since it will have many rows (several thousand rows per session).

Reply:
I meant the line on which I commented, the `uri` field. What is `uri` pointing to?

Reply:
Ah ``uri`` points to the raw data file, e.g., s3://big-ol-data-file.csv. We decided that it was important to be able to track raw data alongside the processed data in the ``Data`` model.